### PR TITLE
Fix free shipping save in setup wizard. (#25139)

### LINKED ISSUE #25139: Free Shipping method is not saved when selected during Setup Wizard
**Describe the bug**

Free Shipping method is not saved when selected during Setup Wizard. 

**To Reproduce**

1. Install and activate WooCommerce on a brand new WordPress installation;
2. Start Setup Wizard;
3. On the Shipping page, select `Free Shipping` for the 2 default zones offered by the wizard:

free_shipping

4. Finish the Setup Wizard;
5. Navigate to `WooCommerce / Settings / Shipping`;
6. You will see that the 2 shipping zones created during the Setup Wizard are there but `Free Shipping` method is not assigned to either of them:

no_free

**Screenshots**

See the screenshots above. 

**Expected behavior**

I expected to see `Free Shipping` method assigned to each shipping zone created during Setup Wizard after completing the Wizard and navigating to `WooCommerce / Settings / Shipping`. 

Note that if you select `Flat Rate` and enter the price for `Flat Rate` during Setup Wizard, you will see the Flat Rate shipping method after the wizard is completed. The issue is only with `Free Shipping` method not being saved. 

**Isolating the problem (mark completed items with an [x]):**
- [x] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [x] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/storefront/).
- [x] I can reproduce this bug consistently using the steps above.

**WordPress Environment**
<details>
```
`
### WordPress Environment ###

WordPress address (URL): https://prior-crane.jurassic.ninja
Site address (URL): https://prior-crane.jurassic.ninja
WC Version: 3.8.1
REST API Version: ✔ 1.0.2
Log Directory Writable: ✔
WP Version: 5.3
WP Multisite: –
WP Memory Limit: 256 MB
WP Debug Mode: ✔
WP Cron: ✔
Language: en_US
External object cache: –

### Server Environment ###

Server Info: Apache/2.4.41 (Unix) OpenSSL/1.0.2g
PHP Version: 7.3.12
PHP Post Max Size: 1 GB
PHP Time Limit: 30
PHP Max Input Vars: 5000
cURL Version: 7.47.0
OpenSSL/1.0.2g

SUHOSIN Installed: –
MySQL Version: 5.7.28-0ubuntu0.16.04.2-log
Max Upload Size: 512 MB
Default Timezone is UTC: ✔
fsockopen/cURL: ✔
SoapClient: ✔
DOMDocument: ✔
GZip: ✔
Multibyte String: ✔
Remote Post: ✔
Remote Get: ✔

### Database ###

WC Database Version: 3.8.1
WC Database Prefix: wp_9757a7fc26_
MaxMind GeoIP Database: ✔
Total Database Size: 3.04MB
Database Data Size: 2.06MB
Database Index Size: 0.98MB
wp_9757a7fc26_woocommerce_sessions: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_9757a7fc26_woocommerce_api_keys: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_9757a7fc26_woocommerce_attribute_taxonomies: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_9757a7fc26_woocommerce_downloadable_product_permissions: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_9757a7fc26_woocommerce_order_items: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_9757a7fc26_woocommerce_order_itemmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_9757a7fc26_woocommerce_tax_rates: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_9757a7fc26_woocommerce_tax_rate_locations: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_9757a7fc26_woocommerce_shipping_zones: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_9757a7fc26_woocommerce_shipping_zone_locations: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_9757a7fc26_woocommerce_shipping_zone_methods: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_9757a7fc26_woocommerce_payment_tokens: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_9757a7fc26_woocommerce_payment_tokenmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_9757a7fc26_woocommerce_log: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_9757a7fc26_commentmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_9757a7fc26_comments: Data: 0.02MB + Index: 0.09MB + Engine InnoDB
wp_9757a7fc26_links: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_9757a7fc26_options: Data: 1.48MB + Index: 0.03MB + Engine InnoDB
wp_9757a7fc26_postmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_9757a7fc26_posts: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_9757a7fc26_termmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_9757a7fc26_terms: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_9757a7fc26_term_relationships: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_9757a7fc26_term_taxonomy: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_9757a7fc26_usermeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_9757a7fc26_users: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_9757a7fc26_wc_download_log: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_9757a7fc26_wc_product_meta_lookup: Data: 0.02MB + Index: 0.09MB + Engine InnoDB
wp_9757a7fc26_wc_tax_rate_classes: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_9757a7fc26_wc_webhooks: Data: 0.02MB + Index: 0.02MB + Engine InnoDB

### Security ###

Secure connection (HTTPS): ✔
Hide errors from visitors: ❌Error messages should not be shown to visitors.

### Active Plugins (3) ###

Companion Plugin: by Osk – 1.11
Jetpack by WordPress.com: by Automattic – 7.9.1
WooCommerce: by Automattic – 3.8.1

### Inactive Plugins (2) ###

Akismet Anti-Spam: by Automattic – 4.1.3
Hello Dolly: by Matt Mullenweg – 1.7.2

### Settings ###

API Enabled: –
Force SSL: –
Currency: GBP (£)
Currency Position: left
Thousand Separator: ,
Decimal Separator: .
Number of Decimals: 2
Taxonomies: Product Types: external (external)
grouped (grouped)
simple (simple)
variable (variable)

Taxonomies: Product Visibility: exclude-from-catalog (exclude-from-catalog)
exclude-from-search (exclude-from-search)
featured (featured)
outofstock (outofstock)
rated-1 (rated-1)
rated-2 (rated-2)
rated-3 (rated-3)
rated-4 (rated-4)
rated-5 (rated-5)

Connected to WooCommerce.com: –

### WC Pages ###

Shop base: #6 - /shop/
Cart: #7 - /cart/
Checkout: #8 - /checkout/
My account: #9 - /my-account/
Terms and conditions: ❌ Page not set

### Theme ###

Name: Twenty Twenty
Version: 1.0
Author URL: https://wordpress.org/
Child Theme: ❌ – If you are modifying WooCommerce on a parent theme that you did not build personally we recommend using a child theme. See: How to create a child theme
WooCommerce Support: ✔

### Templates ###

Overrides: –

### Action Scheduler ###

Complete: 0
Oldest: –
Newest: –

Pending: 0
Oldest: –
Newest: –

Canceled: 0
Oldest: –
Newest: –

In-progress: 0
Oldest: –
Newest: –

Failed: 0
Oldest: –
Newest: –

`
```
</details>

### PR DESCRIPTION
**Pull request by OlegApanovich**
### All Submissions:

* [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

There is some wrong setup wizard options in shipping section. When we try so save shipping option we also save shipping cost settings any way it is a Free Shipping or Flat Rate shipping methods.
Obviously in case of Free Shipping cost option is unnecessary.

Our request $_POST['shipping_zones'] has such value when we try to save it with Free Shipping for a domestic method and with Flate rate for Other locations 

```
  ["domestic"]=>
  array(3) {
    ["method"]=>
    string(13) "free_shipping"
    ["flat_rate"]=>
    array(1) {
      ["cost"]=>
      string(3) "100"
    }
    ["enabled"]=>
    string(3) "yes"
  }
  ["intl"]=>
  array(3) {
    ["method"]=>
    string(9) "flat_rate"
    ["flat_rate"]=>
    array(1) {
      ["cost"]=>
      string(4) "1000"
    }
    ["enabled"]=>
    string(3) "yes"
  }

```
As we can see flat_rate setting with cost 100 is redundant for domestic method, but we have it in our Post request anyway. More over we also try to do checking with it wrong setting.
And it produce fold for a Free Shing method and as result do not save in wocommerce settings after setup wizard over. In my PL I replased checking of all WP_REST_Request request with just individual setting checking,

We also can add additional code to ajax request that trying to remove redundant cost setting for a free shipping but it will become our js code more complicated. For now it quite enough that I added in PL to fix current issue.

Closes #25139 .

### How to test the changes in this Pull Request:

1. All steps already has initial issue.

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [x] Have you successfully run tests with your changes locally?

### Changelog entry
Fix free shipping save in setup wizard.
> Enter a summary of all changes on this Pull Request. This will appear in the changelog if accepted.

### CODE REVIEW

File: includes/admin/class-wc-admin-setup-wizard.php [hunk 1]
```diff
@@ -1276,39 +1276,42 @@ public function wc_setup_shipping_save() {
...
-				$request->set_body(
-					wp_json_encode(
-						array(
-							'method_id' => $domestic_method,
-							'settings'  => wc_clean( wp_unslash( $_POST['shipping_zones']['domestic'][ $domestic_method ] ) ),
-						)
-					)
-				);
-				rest_do_request( $request );
+				$settings = $_POST['shipping_zones']['domestic'][ $domestic_method ];
```
Comment (claudiosanches) [inline includes/admin/class-wc-admin-setup-wizard.php]: Better keep the sanitization here.

File: includes/admin/class-wc-admin-setup-wizard.php [hunk 2]
```diff
@@ -1276,39 +1276,42 @@ public function wc_setup_shipping_save() {
...
-				$request->set_body(
-					wp_json_encode(
-						array(
-							'method_id' => $intl_method,
-							'settings'  => wc_clean( wp_unslash( $_POST['shipping_zones']['intl'][ $intl_method ] ) ),
-						)
-					)
-				);
-				rest_do_request( $request );
+				$settings = $_POST['shipping_zones']['intl'][ $intl_method ];
```
Comment (claudiosanches) [inline includes/admin/class-wc-admin-setup-wizard.php]: Missing sanitization here too.

**Review by claudiosanches — CHANGES_REQUESTED** [review]
All input sanitization got removed in this PR.
Please introduce sanitization again.
Thanks.

**Review by claudiosanches — CHANGES_REQUESTED** [review]
There's some conflicts to solve.

**Comment by OlegApanovich** [comment]
> There's some conflicts to solve.

Conflicts resolved.

**Comment by claudiosanches** [comment]
Thank you very much for your help, but since we are using a new OBW this PR it's no longer necessary, we are going to remove this setup wizard in the next release.

**Pull request closed without merging**